### PR TITLE
Allow compiling async functions to generators

### LINKED ISSUE #1780: Add support for transforming async to a generator
Since generators are available in Chrome, Firefox and io.js it would be nice to transform async functions to a wrapped generator.

### PR DESCRIPTION
**Pull request by arv**
This is done by setting `--async-functions=generator --generators=parse`.

Related to #1231
Fixes #1780

### CODE REVIEW

**Comment by arv** [comment]
@johnjbarton The feature tests now run all `test/feature/AsyncFunctions` twice. Once with whatever options are set in the prolog and once where we override the options to transform to generators. The code for this is pretty ugly. Suggestions welcome.

File: src/codegeneration/AsyncToGeneratorTransformer.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+function wrapBodyInSpawn(body) {
+  const visitor = new FindArguments();
+  visitor.visitAny(body);
+  const argExpr = visitor.found ?
+      createIdentifierExpression(ARGUMENTS) :
+      createNullLiteral();
+  const statement = parseStatement
+      `return $traceurRuntime.spawn(this, ${argExpr}, function*() { ${body} });`
```
Comment (johnjbarton) [inline src/codegeneration/AsyncToGeneratorTransformer.js]: I'm surprised to see es6-ish here. Is there no combination of options where this generator function parse will fail?

File: src/codegeneration/FromOptionsTransformer.js [hunk 2]
```diff
@@ -200,8 +201,16 @@ export class FromOptionsTransformer extends MultiTransformer {
...
+        && !transformOptions.generators) {
+      append(AsyncToGeneratorTransformer);
+    }
+
     // generator must come after for of, for on and rest parameters
-    if (transformOptions.generators || transformOptions.asyncFunctions)
+    if (transformOptions.generators ||
+        transformOptions.asyncFunctions &&
+        transformOptions.asyncFunctions !== 'generator')
       append(GeneratorTransformPass);
```
Comment (johnjbarton) [inline src/codegeneration/FromOptionsTransformer.js]: Oh, I see I should have read a bit further down ;-)

File: test/featureTestRunner.js [hunk 3]
```diff
@@ -121,19 +130,27 @@ function setOptions(load, prologOptions) {
...
-function featureTest(name, url) {
+function featureTest(name, url, overrideOptions = undefined) {
   test(name, (done) => {
     let baseURL = './';
 
     let prologOptions;
     function translateSynchronous(load) {
       let source = load.source;
+
       // Only top level file can set prologOptions.
```
Comment (johnjbarton) [inline test/featureTestRunner.js]: I think this comment is unclear or at least incomplete. Here would be a good spot to explain the two pass testing.

**Comment by johnjbarton** [comment]
Regarding the two-pass tests, duplicating the files with different options is not a horrible alternative and probably makes the system simpler. But it's ok this way.

LGTM

**Comment by arv** [comment]
There is also one more complication and that is that we need to detect if there is native support for generators. For example Node 0.10 does not have generators. I thought of doing some optional skip directive fur the test runner...

File: src/codegeneration/AsyncToGeneratorTransformer.js (hunk 1 above)
Comment (arv) [inline src/codegeneration/AsyncToGeneratorTransformer.js]: No. This is only enabled if parsing generators is enabled.

File: test/featureTestRunner.js (hunk 3 above)
Comment (arv) [inline test/featureTestRunner.js]: This is actually an old comment. It comes from the case where the main test uses `import`, which means that we hit the `translateSynchronous` for the imported files. I'll update the comment.

**Comment by arv** [comment]
I ended up duplicating the test and adding support for conditional skip to the feature test runner. I also changed the options slightly, so that if `--generator=parse` then we generate code that contains generators.

**Pull request merged**